cmd/ipfs/dist/install.sh

```diff
@@ -2,20 +2,86 @@
 
 bin=ipfs
 
-# this script is currently brain dead.
-# it merely tries two locations.
-# in the future maybe use value of $PATH.
-
-binpath=/usr/local/bin
-if [ -d "$binpath" ]; then
-  mv "$bin" "$binpath/$bin"
-  echo "installed $binpath/$bin"
-  exit 0
+old_IFS="$IFS"
+IFS=':'
+
+echo "Welcome to the $bin installer."
+echo "With your permission it will try to install the $bin binary into one of"
+echo "the folders found in your \$PATH variable."
+echo
+
+if [ ! -f "$bin" ]; then
+  echo "The '$bin' binary was not found in the current directory."
+  echo "There is nothing to install."
+  exit 1
+fi
+
+# Make sure that /usr/local/bin and /usr/bin are the first paths we try.
+# Note that these paths might be already found in $PATH.
+paths="/usr/local/bin:/usr/bin:$PATH"
+bestpath=''
+
+echo "We will now try to determine the best directory to install $bin into:"
+echo
+for binpath in $paths;
+do
+  if [ ! -d "$binpath" ]; then
+    echo "$binpath does not seem to be a directory. Skipping..."
+    continue
+  fi
+  if [ ! -w "$binpath" ]; then
+    echo "$binpath does not seem to be writable. Skipping..."
+    continue
+  fi
+  echo -n "$binpath looks like it might work"
+  if [ -z "$bestpath" ]; then
+    bestpath="$binpath"
+    echo ". Let's take it!"
+  else
+    echo " but we already found a better path. Skipping..."
+  fi
+done
+
+# If the user is not root do let them know that they might want to run
+# this script as a privileged user.
+if [ "$(id -u)" != "0" ]; then
+  echo
+  echo "(Note that your might want to run this script as a privileged user,"
+  echo " many of the directories above might become writable.)"
 fi
 
-binpath=/usr/bin
-if [ -d "$binpath" ]; then
-  mv "$bin" "$binpath/$bin"
-  echo "installed $binpath/$bin"
+if [ -z "$bestpath" ]; then
+  echo "No useful path was found."
+  exit 1
+fi
+
+binpath="$bestpath"
+
+echo
+echo "$bin will be installed to '$binpath'"
+while true; do
+    read -p "Is that okay? [Y/n] " yn
+
+    # Default to yes on empty response
+    if [ -z "$yn" ]; then
+      yn="Y"
+    fi
+
+    case $yn in
+        [Yy]* ) break;;
+        [Nn]* ) exit 0;;
+        * ) echo "Please answer yes or no.";;
+    esac
+done
+
+if mv "$bin" "$binpath"; then
```

Comment:
Ah yep
On Thu, Sep 8, 2016 at 10:02 AM Stephan Kulla <notifications@github.com>
wrote:

> In cmd/ipfs/dist/install.sh
> <https://github.com/ipfs/go-ipfs/pull/2504#discussion_r77961445>:
>
> > +while true; do
> > +    read -p "Is that okay? [Y/n] " yn
> > +
> > +    # Default to yes on empty response
> > +    if [ -z "$yn" ]; then
> > +      yn="Y"
> > +    fi
> > +
> > +    case $yn in
> > +        [Yy]* ) break;;
> > +        [Nn]* ) exit 0;;
> > +        * ) echo "Please answer yes or no.";;
> > +    esac
> > +done
> > +
> > +if mv "$bin" "$binpath"; then
>
> mv "$bin" "$binpath/$bin" also has his (very rare) pitfalls. If
> /usr/local/bin/ipfs is a directory, then for binpath=/usr/local/bin and
> bin=ipfs the file ipfs will be stored as /usr/local/bin/ipfs/ipfs.
> Therefore I guess mv -t "$binpath" "$bin" is the better command...
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ipfs/go-ipfs/pull/2504/files/c26ebbcf757b4a5e30f467b6db6849a2c4a81263#r77961445>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAIcoYrcGry4L2APQNt7TfytSPz7nTimks5qn8ERgaJpZM4H42aR>
> .
>